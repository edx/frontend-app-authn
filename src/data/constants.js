// URL Paths
export const LOGIN_PAGE = '/login';
export const REGISTER_PAGE = '/register';
export const RESET_PAGE = '/reset';
export const DEFAULT_REDIRECT_URL = '/dashboard';
export const PASSWORD_RESET_CONFIRM = '/password_reset_confirm/:token/';
export const PAGE_NOT_FOUND = '/notfound';

// Stateful Submit Button States
export const DEFAULT_STATE = 'default';
export const PENDING_STATE = 'pending';

// Constants
export const SUPPORTED_ICON_CLASSES = ['apple', 'facebook', 'google', 'microsoft'];
export const REGISTRATION_VALIDITY_MAP = {};
export const REGISTRATION_OPTIONAL_MAP = {};
export const REGISTRATION_EXTRA_FIELDS = [
  'confirm_email',
  'level_of_education',
  'gender',
  'year_of_birth',
  'mailing_address',
  'goals',
  'honor_code',
  'terms_of_service',
  'city',
  'country',
];

export const VALID_EMAIL_REGEX = '(^[-!#$%&\'*+/=?^_`{}|~0-9A-Z]+(\\.[-!#$%&\'*+/=?^_`{}|~0-9A-Z]+)*'
                                 + '|^"([\\001-\\010\\013\\014\\016-\\037!#-\\[\\]-\\177]|\\\\[\\001-\\011\\013\\014\\016-\\177])*"'
                                 + ')@((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\\.)+)(?:[A-Z0-9-]{2,63})'
                                 + '|\\[(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)(\\.(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)){3}\\]$';
