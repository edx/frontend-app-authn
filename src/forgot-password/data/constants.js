export default 'internal-server-error';
