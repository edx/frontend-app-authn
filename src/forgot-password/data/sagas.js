import { call, put, takeEvery } from 'redux-saga/effects';

// Actions
import {
  FORGOT_PASSWORD,
  forgotPasswordBegin,
  forgotPasswordSuccess,
  forgotPasswordForbidden,
} from './actions';

import { forgotPassword } from './service';

// Services
export function* handleForgotPassword(action) {
  try {
    yield put(forgotPasswordBegin());

    yield call(forgotPassword, action.payload.email);

    yield put(forgotPasswordSuccess(action.payload.email));
  } catch (e) {
    if (e.response && e.response.status === 403) {
      yield put(forgotPasswordForbidden());
    } else {
      throw e;
    }
  }
}

export default function* saga() {
  yield takeEvery(FORGOT_PASSWORD.BASE, handleForgotPassword);
}
