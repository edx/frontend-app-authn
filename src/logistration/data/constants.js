export const INACTIVE_USER = 'inactive-user';
export const INTERNAL_SERVER_ERROR = 'internal-server-error';
export const NON_COMPLIANT_PASSWORD_EXCEPTION = 'NonCompliantPasswordException';
