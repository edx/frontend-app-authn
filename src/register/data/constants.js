export const YEAR_OF_BIRTH_OPTIONS = (() => {
  const currentYear = new Date().getFullYear();
  const years = [];
  let startYear = currentYear - 120;
  while (startYear < currentYear) {
    startYear += 1;

    years.push({ value: startYear.toString(), label: startYear });
  }
  return years.reverse();
})();

export const EDUCATION_LEVELS = [
  '',
  'p',
  'm',
  'b',
  'a',
  'hs',
  'jhs',
  'el',
  'none',
  'o',
];

export const GENDER_OPTIONS = ['', 'f', 'm', 'o'];
