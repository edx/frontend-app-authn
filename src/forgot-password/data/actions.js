import { AsyncActionType } from '../../data/utils';

export const FORGOT_PASSWORD = new AsyncActionType('FORGOT', 'PASSWORD');

// Forgot Password
export const forgotPassword = email => ({
  type: FORGOT_PASSWORD.BASE,
  payload: { email },
});

export const forgotPasswordBegin = () => ({
  type: FORGOT_PASSWORD.BEGIN,
});

export const forgotPasswordSuccess = email => ({
  type: FORGOT_PASSWORD.SUCCESS,
  payload: { email },
});

export const forgotPasswordForbidden = (errorCode) => ({
  type: FORGOT_PASSWORD.FORBIDDEN,
  payload: { errorCode },
});
