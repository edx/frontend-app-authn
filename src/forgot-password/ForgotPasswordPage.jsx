import React, { useState } from 'react';

import { Formik } from 'formik';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

import { getConfig } from '@edx/frontend-platform';
import { sendPageEvent } from '@edx/frontend-platform/analytics';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import {
  Alert,
  Form,
  StatefulButton,
} from '@edx/paragon';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

import { forgotPassword } from './data/actions';
import { forgotPasswordResultSelector } from './data/selectors';
import RequestInProgressAlert from './RequestInProgressAlert';

import messages from './messages';
import {
  AuthnValidationFormGroup,
} from '../common-components';
import APIFailureMessage from '../common-components/APIFailureMessage';
import { INTERNAL_SERVER_ERROR, LOGIN_PAGE, VALID_EMAIL_REGEX } from '../data/constants';
import LoginHelpLinks from '../login/LoginHelpLinks';

const ForgotPasswordPage = (props) => {
  const { intl, status } = props;

  const platformName = getConfig().SITE_NAME;
  const regex = new RegExp(VALID_EMAIL_REGEX, 'i');
  const [validationError, setValidationError] = useState('');

  const getErrorMessage = (errors) => {
    const header = intl.formatMessage(messages['forgot.password.request.server.error']);
    if (errors.email) {
      return (
        <Alert variant="danger">
          <Alert.Heading>{header}</Alert.Heading>
          <ul><li>{errors.email}</li></ul>
        </Alert>
      );
    }
    if (status === INTERNAL_SERVER_ERROR) {
      return <APIFailureMessage header={header} />;
    }
    return status === 'forbidden' ? <RequestInProgressAlert /> : null;
  };

  const getValidationMessage = (email) => {
    let error = '';

    if (email === '') {
      error = intl.formatMessage(messages['forgot.password.empty.email.field.error']);
    } else if (!regex.test(email)) {
      error = intl.formatMessage(messages['forgot.password.page.invalid.email.message']);
    }

    setValidationError(error);
    return error;
  };

  sendPageEvent('login_and_registration', 'reset');

  return (
    <Formik
      initialValues={{ email: '' }}
      validateOnChange={false}
      validate={(values) => {
        const validationMessage = getValidationMessage(values.email);

        if (validationMessage !== '') {
          window.scrollTo({ left: 0, top: 0, behavior: 'smooth' });
          return { email: validationMessage };
        }

        return {};
      }}
      onSubmit={(values) => { props.forgotPassword(values.email); }}
    >
      {({
        errors, handleSubmit, setFieldValue, values,
      }) => (
        <>
          {status === 'complete' ? <Redirect to={LOGIN_PAGE} /> : null}
          <div className="d-flex justify-content-center m-4">
            <div className="d-flex flex-column">
              <Form className="mw-500">
                { getErrorMessage(errors) }
                <h3 className="mt-3">
                  {intl.formatMessage(messages['forgot.password.page.heading'])}
                </h3>
                <p className="mb-4">
                  {intl.formatMessage(messages['forgot.password.page.instructions'])}
                </p>
                <AuthnValidationFormGroup
                  label={intl.formatMessage(messages['forgot.password.page.email.field.label'])}
                  for="forgot-password-input"
                  name="email"
                  type="email"
                  invalid={validationError !== ''}
                  invalidMessage={validationError}
                  placeholder="username@domain.com"
                  value={values.email}
                  onBlur={() => getValidationMessage(values.email)}
                  onChange={e => setFieldValue('email', e.target.value)}
                  helpText={intl.formatMessage(messages['forgot.password.email.help.text'], { platformName })}
                  className="mb-0 w-100"
                />
                <LoginHelpLinks page="forgot-password" />
                <StatefulButton
                  type="submit"
                  className="btn-primary mt-3"
                  state={status}
                  labels={{
                    default: intl.formatMessage(messages['forgot.password.page.submit.button']),
                  }}
                  icons={{ pending: <FontAwesomeIcon icon={faSpinner} spin /> }}
                  onClick={handleSubmit}
                  onMouseDown={(e) => e.preventDefault()}
                />
              </Form>
            </div>
          </div>
        </>
      )}
    </Formik>
  );
};

ForgotPasswordPage.propTypes = {
  intl: intlShape.isRequired,
  forgotPassword: PropTypes.func.isRequired,
  status: PropTypes.string,
};

ForgotPasswordPage.defaultProps = {
  status: null,
};

export default connect(
  forgotPasswordResultSelector,
  {
    forgotPassword,
  },
)(injectIntl(ForgotPasswordPage));
