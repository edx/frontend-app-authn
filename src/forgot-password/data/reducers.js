import { FORGOT_PASSWORD } from './actions';

export const defaultState = {
  status: null,
};

const reducer = (state = defaultState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case FORGOT_PASSWORD.BEGIN:
        return {
          ...state,
          status: 'pending',
        };
      case FORGOT_PASSWORD.SUCCESS:
        return {
          ...state,
          ...action.payload,
          status: 'complete',
        };
      case FORGOT_PASSWORD.FORBIDDEN:
        return {
          ...state,
          status: 'forbidden',
          errorCode: action.payload.errorCode,
        };
      default:
        return state;
    }
  }
  return state;
};

export default reducer;
