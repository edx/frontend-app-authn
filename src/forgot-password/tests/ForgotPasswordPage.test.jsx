import React from 'react';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { IntlProvider, injectIntl } from '@edx/frontend-platform/i18n';
import CookiePolicyBanner from '@edx/frontend-component-cookie-policy-banner';
import * as analytics from '@edx/frontend-platform/analytics';

import ForgotPasswordPage from '../ForgotPasswordPage';
import { INTERNAL_SERVER_ERROR } from '../../data/constants';

jest.mock('@edx/frontend-platform/analytics');

analytics.sendPageEvent = jest.fn();

const IntlForgotPasswordPage = injectIntl(ForgotPasswordPage);
const mockStore = configureStore();
const initialState = {
  forgotPassword: {
    status: null,
  },
};

describe('ForgotPasswordPage', () => {
  let props = {};
  let store = {};

  const reduxWrapper = children => (
    <IntlProvider locale="en">
      <MemoryRouter>
        <Provider store={store}>{children}</Provider>
      </MemoryRouter>
    </IntlProvider>
  );

  beforeEach(() => {
    store = mockStore(initialState);
    props = {
      forgotPassword: jest.fn(),
      status: null,
    };
  });

  it('should match default section snapshot', () => {
    const tree = renderer.create(reduxWrapper(<IntlForgotPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match forbidden section snapshot', () => {
    props = {
      ...props,
      status: 'forbidden',
    };
    const tree = renderer.create(reduxWrapper(<IntlForgotPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match pending section snapshot', () => {
    props = {
      ...props,
      status: 'pending',
    };
    const tree = renderer.create(reduxWrapper(<IntlForgotPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display need other help signing in button', () => {
    const wrapper = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));
    expect(wrapper.find('#forgot-password.btn-link').first().text()).toEqual('Need help signing in?');
  });

  it('should display email validation error message', async () => {
    const validationMessage = 'We were unable to contact you.Enter a valid email address below.';
    const wrapper = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));

    wrapper.find('input#email').simulate(
      'change', { target: { value: 'invalid-email', name: 'email' } },
    );
    await act(async () => { await wrapper.find('button.btn-brand').simulate('click'); });
    wrapper.update();

    expect(wrapper.find('.alert-danger').text()).toEqual(validationMessage);
  });

  it('should show alert on server error', () => {
    store = mockStore({
      forgotPassword: { status: INTERNAL_SERVER_ERROR },
    });
    const expectedMessage = 'We were unable to contact you.'
                            + 'An error has occurred. Try refreshing the page, or check your internet connection.';
    const wrapper = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));

    expect(wrapper.find('#internal-server-error').first().text()).toEqual(expectedMessage);
  });

  it('should display empty email validation message', async () => {
    const validationMessage = 'We were unable to contact you.Enter your email below.';
    const forgotPasswordPage = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));

    await act(async () => { await forgotPasswordPage.find('button.btn-brand').simulate('click'); });

    forgotPasswordPage.update();
    expect(forgotPasswordPage.find('.alert-danger').text()).toEqual(validationMessage);
  });

  it('should display request in progress error message', () => {
    const rateLimitMessage = 'An error occurred.Your previous request is in progress, please try again in a few moments.';
    store = mockStore({
      forgotPassword: { status: 'forbidden' },
    });

    const forgotPasswordPage = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));
    expect(forgotPasswordPage.find('.alert-danger').text()).toEqual(rateLimitMessage);
  });

  it('should not display any error message on change event', () => {
    const forgotPasswordPage = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));

    const emailInput = forgotPasswordPage.find('input#email');
    emailInput.simulate('change', { target: { value: 'invalid-email', name: 'email' } });
    forgotPasswordPage.update();

    expect(forgotPasswordPage.find('#email-invalid-feedback').exists()).toEqual(false);
  });

  it('should display error message on blur event', async () => {
    const validationMessage = 'Enter your email';
    const forgotPasswordPage = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));
    const emailInput = forgotPasswordPage.find('input#email');

    await act(async () => {
      await emailInput.simulate('blur', { target: { value: '', name: 'email' } });
    });

    forgotPasswordPage.update();
    expect(forgotPasswordPage.find('.pgn__form-text-invalid').text()).toEqual(validationMessage);
  });

  it('check cookie rendered', () => {
    const forgotPage = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));
    expect(forgotPage.find(<CookiePolicyBanner />)).toBeTruthy();
  });

  it('should display success message after email is sent', async () => {
    store = mockStore({
      ...initialState,
      forgotPassword: {
        status: 'complete',
      },
    });
    const successMessage = 'Check your emailWe sent an email to  with instructions to reset your password. If you do not '
                           + 'receive a password reset message after 1 minute, verify that you entered the correct email address,'
                           + ' or check your spam folder. If you need further assistance, contact technical support.';

    const wrapper = mount(reduxWrapper(<IntlForgotPasswordPage {...props} />));
    expect(wrapper.find('.alert-success').text()).toEqual(successMessage);
  });
});
