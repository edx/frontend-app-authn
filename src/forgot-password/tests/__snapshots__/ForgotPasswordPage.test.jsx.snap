// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPasswordPage should match default section snapshot 1`] = `
<div
  className="d-flex justify-content-center m-4"
>
  <div
    className="d-flex flex-column"
  >
    <form
      className="mw-500"
    >
      <h3
        className="mt-3"
      >
        Password assistance
      </h3>
      <p
        className="mb-4"
      >
        Please enter your log-in or recovery email address below and we will send you an email with instructions.
      </p>
      <div
        className="form-group mb-0 w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="forgot-password-input"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          required={true}
          type="email"
          value=""
        />
        <span />
      </div>
      <button
        className="mt-2 field-link small"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-caret-right fa-w-6 mr-1"
          data-icon="caret-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Need other help signing in?
      </button>
      <div
        className="pgn-transition-replace-group position-relative"
        style={
          Object {
            "height": null,
          }
        }
      >
        <div
          style={
            Object {
              "padding": ".1px 0",
            }
          }
        />
      </div>
      <button
        aria-disabled={false}
        aria-live="assertive"
        className="pgn__stateful-btn pgn__stateful-btn-state-null btn-primary mt-3 btn btn-primary"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        type="submit"
      >
        <span
          className="d-flex align-items-center justify-content-center"
        >
          Recover my password
        </span>
      </button>
    </form>
  </div>
</div>
`;

exports[`ForgotPasswordPage should match forbidden section snapshot 1`] = `
<div
  className="d-flex justify-content-center m-4"
>
  <div
    className="d-flex flex-column"
  >
    <form
      className="mw-500"
    >
      <div
        className="fade alert alert-danger show"
        role="alert"
      >
        <div
          className="alert-heading h4"
        >
          An error occurred.
        </div>
        <ul>
          <li>
            Your previous request is in progress, please try again in a few moments.
          </li>
        </ul>
      </div>
      <h3
        className="mt-3"
      >
        Password assistance
      </h3>
      <p
        className="mb-4"
      >
        Please enter your log-in or recovery email address below and we will send you an email with instructions.
      </p>
      <div
        className="form-group mb-0 w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="forgot-password-input"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          required={true}
          type="email"
          value=""
        />
        <span />
      </div>
      <button
        className="mt-2 field-link small"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-caret-right fa-w-6 mr-1"
          data-icon="caret-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Need other help signing in?
      </button>
      <div
        className="pgn-transition-replace-group position-relative"
        style={
          Object {
            "height": null,
          }
        }
      >
        <div
          style={
            Object {
              "padding": ".1px 0",
            }
          }
        />
      </div>
      <button
        aria-disabled={false}
        aria-live="assertive"
        className="pgn__stateful-btn pgn__stateful-btn-state-forbidden btn-primary mt-3 btn btn-primary"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        type="submit"
      >
        <span
          className="d-flex align-items-center justify-content-center"
        >
          Recover my password
        </span>
      </button>
    </form>
  </div>
</div>
`;

exports[`ForgotPasswordPage should match pending section snapshot 1`] = `
<div
  className="d-flex justify-content-center m-4"
>
  <div
    className="d-flex flex-column"
  >
    <form
      className="mw-500"
    >
      <h3
        className="mt-3"
      >
        Password assistance
      </h3>
      <p
        className="mb-4"
      >
        Please enter your log-in or recovery email address below and we will send you an email with instructions.
      </p>
      <div
        className="form-group mb-0 w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="forgot-password-input"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          required={true}
          type="email"
          value=""
        />
        <span />
      </div>
      <button
        className="mt-2 field-link small"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-caret-right fa-w-6 mr-1"
          data-icon="caret-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Need other help signing in?
      </button>
      <div
        className="pgn-transition-replace-group position-relative"
        style={
          Object {
            "height": null,
          }
        }
      >
        <div
          style={
            Object {
              "padding": ".1px 0",
            }
          }
        />
      </div>
      <button
        aria-disabled={true}
        aria-live="assertive"
        className="pgn__stateful-btn pgn__stateful-btn-state-pending btn-primary mt-3 disabled btn btn-primary"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        type="submit"
      >
        <span
          className="d-flex align-items-center justify-content-center"
        >
          <span
            className="pgn__stateful-btn-icon"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-spinner fa-w-16 fa-spin "
              data-icon="spinner"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </span>
          Recover my password
        </span>
      </button>
    </form>
  </div>
</div>
`;
